File: aoc-2024/java/Day3.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Day3 {
  public static void main(String[] args) throws FileNotFoundException {
    File file = new File("./input/day3.txt");
    Scanner scan = new Scanner(file);
    Pattern pattern = Pattern.compile("(mul\\([0-9]*,[0-9]*\\)|don't|do?)");
    int total = 0;
    boolean enabled = true;
    while (scan.hasNextLine()) {
      String line = scan.nextLine();
      Matcher matcher = pattern.matcher(line);
      while (matcher.find()) {
        String match = matcher.group();
        if (match.equals("do")) {
          enabled = true;
        } else if (match.equals("don't")) {
          enabled = false;
        } else if (enabled) {
          String[] mults = match.replace("mul(", "").replace(")", "").split(",");
          int result = Integer.parseInt(mults[0]) * Integer.parseInt(mults[1]);
          total += result;
          System.out.println(match + ", "  + result);
        }
      }
    }
    System.out.println(total);
  }
}

File: aoc-2024/day3/day3.py
import re

lines = open('./day3.txt')
total = 0
enabled = True
for line in lines:
  matches = re.findall("(mul\([0-9]*,[0-9]*\)|don't|do?)", line)
  for match in matches:  
    if match == "don't":
      enabled = False
    elif match == 'do':
      enabled = True
    elif enabled:
      mults = [int(x) for x in match.lstrip("mul(").rstrip(")").split(",")]
      total += mults[0] * mults[1]
      print(mults, total)
print(total)